File: app/src/main/java/com/spacebitlabs/plantly/plants/ToolbarBehavior.java
package com.spacebitlabs.plantly.plants;

import android.content.Context;
import android.support.annotation.Keep;
import android.support.annotation.NonNull;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.v4.view.ViewCompat;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import timber.log.Timber;

/**
 * Toolbar Behavior to resize CircleImageViews
 * while scrolling
 */

@Keep // might be needed for minification
public class ToolbarBehavior extends CoordinatorLayout.Behavior<LinearLayout> {

    private float maxScrollValue;
    private float initialY;
    private int initialImageWidth;
    private int initialImageHeight;

    public ToolbarBehavior(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    public boolean layoutDependsOn(CoordinatorLayout parent, LinearLayout child, View dependency) {
        return dependency instanceof AppBarLayout;
    }

    @Override
    public boolean onStartNestedScroll(@NonNull CoordinatorLayout coordinatorLayout, @NonNull LinearLayout child, @NonNull View directTargetChild, @NonNull View target, int axes, int type) {
        return (axes & ViewCompat.SCROLL_AXIS_VERTICAL) != 0;
    }

    @Override
    public boolean onDependentViewChanged(CoordinatorLayout parent, LinearLayout child, View dependency) {
        if (((ViewGroup) child.getChildAt(2)).getChildAt(0) == null) {
            return false;
        }
        maybeInitProperties(child, dependency);

        float scrollValuePercentage = 1 - (maxScrollValue + dependency.getY()) / maxScrollValue;
        scrollValuePercentage = Math.min(1.5f * scrollValuePercentage, 1); // speed up the animations relative to the movement

//        Timber.d("maxScrollValue: " + scrollValuePercentage);

        child.setY(initialY + dependency.getY());

        ViewGroup todayListView = (ViewGroup) getScrollView(child);
        int childCount = todayListView.getChildCount();

        // resize all the images
        for (int i = 0; i < childCount; i++) {
            ViewGroup todayListViewItem = (ViewGroup) todayListView.getChildAt(i);
            View circleImageView = todayListViewItem.getChildAt(0);
            LinearLayout.LayoutParams lp1 = (LinearLayout.LayoutParams) circleImageView.getLayoutParams();
            lp1.width = (int) (initialImageWidth * Math.min(1, 1 - scrollValuePercentage + 0.5));
            lp1.height = (int) (initialImageHeight * Math.min(1, 1 - scrollValuePercentage + 0.5));

            circleImageView.setLayoutParams(lp1);
        }

        child.setAlpha(1 - scrollValuePercentage);

        return true;
    }

    private void maybeInitProperties(LinearLayout child, View dependency) {
        if (maxScrollValue == 0) {
            maxScrollValue = dependency.getHeight() - dependency.getMinimumHeight();
        }

        if (initialY == 0) {
            initialY = child.getY();
        }

        if (initialImageWidth == 0) {
            initialImageWidth = getTodayItemImage(child).getWidth();
        }

        if (initialImageHeight == 0) {
            initialImageHeight = getTodayItemImage(child).getHeight();
        }
    }

    private View getScrollView(LinearLayout child) {
        return child.getChildAt(2);
    }

    private View getTodayItemImage(LinearLayout child) {
        return ((ViewGroup) ((ViewGroup)
          child.getChildAt(2)) // RecyclerView
                               .getChildAt(0)) // LinearLayout
                                               .getChildAt(0); // CircleImageView;
    }
}

File: app/src/main/java/com/spacebitlabs/plantly/plants/ToolbarBehavior.kt
package com.spacebitlabs.plantly.plants

import android.content.Context
import com.google.android.material.appbar.AppBarLayout
import androidx.coordinatorlayout.widget.CoordinatorLayout
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import kotlinx.android.synthetic.main.fragment_plants.view.*
import kotlinx.android.synthetic.main.today_list_item.view.*

@Suppress("unused")
class ToolbarBehavior(context: Context, attrs: AttributeSet) :
    androidx.coordinatorlayout.widget.CoordinatorLayout.Behavior<LinearLayout>(context, attrs) {

    private var initialised = false
    private var maxScrollValue = 0f
    private var initialImageWidth = 0f
    private var initialImageHeight = 0f

    override fun layoutDependsOn(parent: androidx.coordinatorlayout.widget.CoordinatorLayout, child: LinearLayout, dependency: View): Boolean {
        return dependency is AppBarLayout
    }

    override fun onDependentViewChanged(parent: androidx.coordinatorlayout.widget.CoordinatorLayout, child: LinearLayout, dependency: View): Boolean {
        maybeInitProperties(child, dependency)

        var scrollValuePercentage = 1 - (maxScrollValue + dependency.y) / maxScrollValue
        scrollValuePercentage = Math.min(1.5f * scrollValuePercentage, 1f) // speed up the animations relative to the movement

//        Timber.d("maxScrollValue: %s", scrollValuePercentage);
//        Timber.d("InitialY: %s", initialY);

        child.y = dependency.y
        child.alpha = 1 - scrollValuePercentage

        if (child.today_list.childCount > 0) {

            val todayListView = child.today_list
            val childCount = todayListView.childCount

            // resize all the images
            for (i in 0 until childCount) {
                val todayListViewItem = todayListView.getChildAt(i) as ViewGroup
                val circleImageView = todayListViewItem.getChildAt(0)
                val lp1 = circleImageView.layoutParams as LinearLayout.LayoutParams
                lp1.width = (initialImageWidth * Math.min(1.0, 1 - scrollValuePercentage + 0.5)).toInt()
                lp1.height = (initialImageHeight * Math.min(1.0, 1 - scrollValuePercentage + 0.5)).toInt()

                circleImageView.layoutParams = lp1
            }
        }

        return true
    }

    private fun maybeInitProperties(child: LinearLayout, dependency: View) {
        if (maxScrollValue == 0f) {
            maxScrollValue = (dependency.height - dependency.minimumHeight).toFloat()
        }

        if (child.today_list.childCount > 0 && initialImageWidth == 0f) {
            initialImageWidth = getTodayItemImage(child).width.toFloat()
        }

        if (child.today_list.childCount > 0 && initialImageHeight == 0f) {
            initialImageHeight = getTodayItemImage(child).height.toFloat()
        }
    }

    private fun getTodayItemImage(child: LinearLayout): View = child.today_list.image
}